src/slurm_plugin/clustermgtd.py

```diff
@@ -964,6 +992,95 @@ def _is_node_in_replacement_valid(self, node, check_node_is_valid):
+
+        # Reset nodes which insufficient capacity timeout expired
+        if timeout_expired_compute_resources:
+            self._reset_insufficient_capacity_timeout_expired_nodes(
+                timeout_expired_compute_resources, ice_compute_resources_and_nodes_map
+            )
+
+    def _set_ice_compute_resources_to_down(self, queue_compute_resource_nodes_map: dict):
+        """Set powered_down nodes which belong to insufficient capacity compute resources to down."""
+        if not self._insufficient_capacity_compute_resources:
+            return
+        nodes_to_down = {}
+        for queue_compute_resource_name, event in self._insufficient_capacity_compute_resources.items():
+            nodes = queue_compute_resource_nodes_map.get(queue_compute_resource_name)
+            for node in nodes:
+                if node.is_powered_down_and_not_ice():
+                    error_code = event.error_code
+                    nodes_to_down.setdefault(error_code, []).append(node.name)
+        if nodes_to_down:
+            for error_code, node_list in nodes_to_down.items():
+                log.info(
+                    "Setting following nodes into DOWN state due to insufficient capacity: %s",
+                    print_with_count(node_list),
+                )
+                set_nodes_down(
+                    node_list, reason=f"(Code:{error_code})Temporarily disabling node due to insufficient capacity"
+                )
+
+    def _find_insufficient_capacity_timeout_expired_compute_resources(self) -> set:
+        """Find compute resources which insufficient_capacity_timeout expired."""
+        timeout_expired_cr = set()
+        for cr, event in self._insufficient_capacity_compute_resources.copy().items():
+            if time_is_up(event.timestamp, self._current_time, self._config.insufficient_capacity_timeout):
+                self._insufficient_capacity_compute_resources.pop(cr)
+                timeout_expired_cr.add(cr)
+        return timeout_expired_cr
+
+    def _reset_insufficient_capacity_timeout_expired_nodes(
+        self, timeout_expired_cr: set, ice_compute_resources_and_nodes_map: dict
+    ):
+        """Reset nodes in the compute resource which insufficient_capacity_timeout expired."""
+        logging.info(
+            f"Reset the following compute resources because insufficient capacity timeout expired: {timeout_expired_cr}"
+        )
+        nodes_to_power_down = []
+        for compute_resource, data in ice_compute_resources_and_nodes_map.items():
+            if compute_resource in timeout_expired_cr:
+                nodes_to_power_down += data.nodes
+
+        if nodes_to_power_down:
+            log.info(
+                "Enabling the following nodes because insufficient capacity timeout expired: %s",
+                print_with_count(nodes_to_power_down),
+            )
+            set_nodes_power_down(
+                [node.name for node in nodes_to_power_down],
```

Comment:
you could avoid this additional loop by storing node.name in nodes_to_power_down above

Reply:
Changed to
```
        if nodes_to_power_down:
            node_names = [node.name for node in nodes_to_power_down]
            log.info(
                "Enabling the following nodes because insufficient capacity timeout expired: %s",
                print_with_count(node_names),
            )
            set_nodes_power_down(
                node_names,
                reason="Enabling node since insufficient capacity timeout expired",
            )
```